services/src/main/java/io/grpc/protobuf/service/ProtoReflectionService.java

```diff
@@ -380,4 +283,286 @@ private ServerReflectionResponse createServerReflectionResponse(
           .build();
     }
   }
+
+  /**
+   * Indexes the server's services and allows lookups of file descriptors by filename, symbol, type,
+   * and extension number.
+   *
+   * <p>Internally, this stores separate indices for the immutable and mutable services. When
+   * queried, the immutable service index is checked for a matching value. Only if there is no match
+   * in the immutable service index are the mutable services checked.
+   */
+  private class ServerReflectionIndex {
+    private final FileDescriptorIndex immutableServicesIndex = new FileDescriptorIndex();
+    /** Tracks mutable services. Accesses must be synchronized. */
+    @GuardedBy("mutableServicesIndex") private final FileDescriptorIndex mutableServicesIndex =
+        new FileDescriptorIndex();
+
+    private boolean immutableIndexInitialized = false;
+    @GuardedBy("mutableServicesIndex") private Set<FileDescriptor>
+        cachedMutableServiceFileDescriptors = new HashSet<FileDescriptor>();
+
+    /**
+     * When first called, initializes the immutable services index. Subsequent calls have no effect.
+     *
+     * <p>This must be called by the reflection service before returning a new
+     * {@link ProtoReflectionStreamObserver}.
+     */
+    private synchronized void initializeImmutableServicesIndex() {
+      if (immutableIndexInitialized) {
+        return;
+      }
+      immutableServicesIndex.initialize(server.getImmutableServices());
+      immutableIndexInitialized = true;
+    }
+
+    /**
+     * Checks for updates to the server's mutable services and updates the indices if any changes
+     * are detected. A change is any addition or removal in the set of file descriptors attached to
+     * the mutable services or a change in the service names.
+     */
+    private void updateMutableIndexIfNecessary() {
+      synchronized (mutableServicesIndex) {
+        Set<FileDescriptor> currentMutableServiceFileDescriptors =
+            new HashSet<FileDescriptor>();
+        Set<String> currentMutableServiceNames = new HashSet<String>();
+        for (ServerServiceDefinition mutableService : server.getMutableServices()) {
+          io.grpc.ServiceDescriptor serviceDescriptor = mutableService.getServiceDescriptor();
+          if (serviceDescriptor.getMarshallerDescriptor() instanceof ProtoFileDescriptorSupplier) {
+            FileDescriptor fileDescriptor =
+                ((ProtoFileDescriptorSupplier) serviceDescriptor.getMarshallerDescriptor())
+                    .getFileDescriptor();
+            currentMutableServiceFileDescriptors.add(fileDescriptor);
+            currentMutableServiceNames.add(serviceDescriptor.getName());
+          }
+        }
+
+        if (!cachedMutableServiceFileDescriptors.equals(currentMutableServiceFileDescriptors)) {
+          mutableServicesIndex.initialize(server.getMutableServices());
```

Comment:
Done - now caching the services at start of synchronized block.  If they're already stale by this line, the next call will update them again.